beetsplug/replaygain.py

```diff
@@ -271,6 +284,210 @@ def end_element_handler(name):
             out.append(album_gain["album"])
         return out
 
+    def use_ebu_r128(self):
+        """Set this Backend up to use EBU R128."""
+        self.method = '--ebu'
+
+
+# ffmpeg backend
+class FfmpegBackend(Backend):
+    """A replaygain backend using ffmpegs ebur128 filter.
+    """
+    def __init__(self, config, log):
+        super(FfmpegBackend, self).__init__(config, log)
+        config.add({
+            "peak": "true"
+        })
+        self._peak_method = config["peak"].as_str()
+        # convert db to LUFS according to:
+        # http://wiki.hydrogenaud.io/index.php?title=ReplayGain_
+        #   specification#Reference_level
+        self._target_level = config['targetlevel'].as_number() - 103
```

Comment:
How exactly are the options `tragetlevel` and `r128` supposed to work together?
The only backend that already supports `r128`, `Bs1770gainBackend`, just completely ignores `targetlevel`. (That might be a bug?)

Reply:
In `Bs1770gainBackend` initialising it as an EBU R128 Backend only sets the target level to -23 LUFS. So it is impossible to configure the target level of R128_* tags.
Maybe it would make sense to add a config `r128_targetlevel`, defaulting to `80` db (that corresponds to -23 LUFS according to http://wiki.hydrogenaud.io/index.php?title=ReplayGain_specification#Reference_level).
Then add a parameter `target_level` to `Backend.compute_track_gain` and `Backend.compute_album_gain` that is either `targetlevel` or `r128_targetlevel` depending on the filetype. That also makes the whole `ReplayGainPlugin.r128_backend_instance` obsolete, the same backend can just be called with different target levels.
I guess I should create a separate pull request implementing that change?

Also `Bs1770gainBackend`'s method configuration just changes the target level. Shouldn't that be controlled by `targetlevel`?

Reply:
Yes, I like that proposed design a lot—with that change, it would be great to dispense with the distinction between different "backend instances" for classic RG and R128.

I also agree that it seems wrong for `Bs1770gainBackend` to unilaterally change the target level. It does seem like the configuration should control that.

Reply:
This has been implemented in #3065.